packages/devtools_shared/lib/src/extensions/extension_manager.dart

```diff
@@ -0,0 +1,168 @@
+// Copyright (c) 2023, the Dart project authors.  Please see the AUTHORS file
+// for details. All rights reserved. Use of this source code is governed by a
+// BSD-style license that can be found in the LICENSE file.
+
+import 'dart:io';
+
+import 'package:io/io.dart';
+import 'package:path/path.dart' as path;
+
+import 'extension_model.dart';
+
+/// Location where DevTools extension assets will be served, relative to where
+/// DevTools assets are served (build/).
+const extensionRequestPath = 'devtools_extensions';
+
+/// The location for the DevTools extension, relative to the parent package's
+/// root.
+const extensionLocation = 'extension/devtools';
+
+/// The default location for the DevTools extension, relative to
+/// `<parent_package_root>/extension/devtools/`.
+const extensionBuildDefault = 'build';
+
+class ExtensionsManager {
+  ExtensionsManager({required this.buildDir});
+
+  final String buildDir;
+
+  String get servedExtensionsPath => path.join(buildDir, extensionRequestPath);
+
+  final devtoolsExtensions = <DevToolsExtensionConfig>[];
+
+  void serveAvailableExtensions(String? rootPath) {
+    devtoolsExtensions.clear();
+
+    if (rootPath != null) {
+      // TODO(kenz): use 'findExtensions' from package:extension_discovery once it
+      // is published.
+      // final extensions = findExtensions(
+      //   'devtools',
+      //   packageConfig: '$rootPath/.dart_tool/package_config.json',
+      // );
+      final extensions = <Extension>[];
+      for (final extension in extensions) {
+        final config = extension.config;
+        if (config is! Map) {
```

Comment:
If `config` is always expected to be `Map`, why don't we restrict it to `Map<Object, Object?>` in `Extension`?

Reply:
See discussion at https://github.com/dart-lang/tools/pull/129#discussion_r1269800754